### PR TITLE
Add smartprofiles plugin

### PR DESCRIPTION
**Pull request by drshrey**
Adds:

1. SmartProfiles plugin that queries influx and computes a recommended state for project services
2. InfluxDB vendor dependency that's used in the SmartProfiles plugin
3. AutoscaleEnabled option for project services. Nothing is acted upon it for now, but in the future, we will have services use the "suggested" service spec if AutoscaleEnabled is true.
4. Adds suggested service spec and SmartProfile-related events handler in codeamp plugin

### CODE REVIEW

**Comment by aballman** [comment]
Can you pull the influxdb lib out into another pr. This isn't a small pr on it's own and trying to view the diffs with 2500 files changed total (including the lib) is difficult. If you didn't put it in a clean commit, you could add the lib in a pr and merge that into master then merge master into this to reduce the diffs.

**Comment by drshrey** [comment]
@aballman added influx in this PR: https://github.com/codeamp/circuit/pull/392

File: plugins/codeamp/db/service.go [hunk 1]
```diff
@@ -36,11 +36,16 @@ func (r *ServiceResolver) Name() string {
...
+func (r *ServiceResolver) AutoscaleEnabled() bool {
+	return r.Service.AutoscaleEnabled
+}
+
 // ServiceSpec
 func (r *ServiceResolver) ServiceSpec() *ServiceSpecResolver {
 	var serviceSpec model.ServiceSpec
 
-	if err := r.DB.Where("service_id = ?", r.Service.Model.ID).First(&serviceSpec).Error; err != nil {
+	if err := r.DB.Where("service_id = ? and type != ?", r.Service.Model.ID, "suggested").First(&serviceSpec).Error; err != nil {
```
Comment (aballman) [inline plugins/codeamp/db/service.go]: What's the intent here?

File: plugins/codeamp/db/service.go [hunk 2]
```diff
@@ -50,6 +55,15 @@ func (r *ServiceResolver) ServiceSpec() *ServiceSpecResolver {
 	return &ServiceSpecResolver{DB: r.DB, ServiceSpec: serviceSpec}
 }
 
+// SuggestedServiceSpec
+func (r *ServiceResolver) SuggestedServiceSpec() *ServiceSpecResolver {
+	var serviceSpec model.ServiceSpec
+
+	r.DB.Where("service_id = ? and type = ?", r.Service.Model.ID, "suggested").Order("created_at desc").First(&serviceSpec)
```
Comment (aballman) [inline plugins/codeamp/db/service.go]: Please add logging if there's a failure in the query other than the record is not found

File: plugins/codeamp/graphql/service_mutation.go [hunk 3]
```diff
@@ -22,8 +22,10 @@ func (r *ServiceResolverMutation) CreateService(args *struct{ Service *model.Ser
 		return nil, fmt.Errorf("Service name cannot be longer than 63 characters.")
 	}
 
+	tx := r.DB.Begin()
+
 	// Check if project can create service in environment
-	if r.DB.Where("environment_id = ? and project_id = ?", args.Service.EnvironmentID, args.Service.ProjectID).Find(&model.ProjectEnvironment{}).RecordNotFound() {
+	if tx.Where("environment_id = ? and project_id = ?", args.Service.EnvironmentID, args.Service.ProjectID).Find(&model.ProjectEnvironment{}).RecordNotFound() {
```
Comment (aballman) [inline plugins/codeamp/graphql/service_mutation.go]: Log the error as well as return it

File: plugins/codeamp/graphql/service_mutation.go [hunk 4]
```diff
@@ -104,20 +108,23 @@ func (r *ServiceResolverMutation) CreateService(args *struct{ Service *model.Ser
 		IsDefault: false,
 	}
 
-	r.DB.Create(&serviceSpec)
+	if err := tx.Create(&serviceSpec).Error; err != nil {
```
Comment (aballman) [inline plugins/codeamp/graphql/service_mutation.go]: Log the error

File: plugins/codeamp/graphql/servicespec_mutation.go [hunk 5]
```diff
@@ -117,41 +116,35 @@ func (r *ServiceSpecResolverMutation) UpdateServiceSpec(args *struct{ ServiceSpe
...
-	} else {
-		services := []model.Service{}
-		if err := tx.Where("id = ?", serviceSpec.ServiceID).Find(&services).Error; err != nil {
-			tx.Rollback()
+		return nil, fmt.Errorf("Select another service spec to be a default before deleting this.")
+	}	
+
+	if err := tx.Where("id = ?", serviceSpec.ServiceID).First(&service).Error; gorm.IsRecordNotFoundError(err) {
+		if err := tx.Delete(&serviceSpec).Error; err != nil {
+			tx.Rollback()
```
Comment (aballman) [inline plugins/codeamp/graphql/servicespec_mutation.go]: log

File: plugins/codeamp/graphql/servicespec_mutation.go [hunk 6]
```diff
@@ -117,41 +116,35 @@ func (r *ServiceSpecResolverMutation) UpdateServiceSpec(args *struct{ ServiceSpe
...
+	}
+	
+	if serviceSpec.IsDefault {
 		tx.Rollback()
-		return nil, fmt.Errorf("ServiceSpec not found with given argument id")
-	} else {
-		services := []model.Service{}
-		if err := tx.Where("id = ?", serviceSpec.ServiceID).Find(&services).Error; err != nil {
-			tx.Rollback()
+		return nil, fmt.Errorf("Select another service spec to be a default before deleting this.")
```
Comment (aballman) [inline plugins/codeamp/graphql/servicespec_mutation.go]: log

File: plugins/codeamp/graphql/servicespec_mutation.go [hunk 7]
```diff
@@ -117,41 +116,35 @@ func (r *ServiceSpecResolverMutation) UpdateServiceSpec(args *struct{ ServiceSpe
...
+		if err := tx.Delete(&serviceSpec).Error; err != nil {
+			tx.Rollback()			
 			return nil, err
 		}
 		
-		if serviceSpec.IsDefault {
-			tx.Rollback()
-			return nil, fmt.Errorf("Select another service spec to be a default before deleting this.")
+		if err := tx.Commit().Error; err != nil {
+			tx.Rollback()
```
Comment (aballman) [inline plugins/codeamp/graphql/servicespec_mutation.go]: log

File: plugins/codeamp/graphql/servicespec_mutation.go [hunk 8]
```diff
@@ -117,41 +116,35 @@ func (r *ServiceSpecResolverMutation) UpdateServiceSpec(args *struct{ ServiceSpe
...
-
-			//r.ServiceSpecDeleted(&serviceSpec)
 
-			return &ServiceSpecResolver{DBServiceSpecResolver: &db_resolver.ServiceSpecResolver{DB: r.DB, ServiceSpec: serviceSpec}}, nil
-		} else {
-			return nil, fmt.Errorf("Delete all project-services using this service spec first.")
-		}
+		return &ServiceSpecResolver{DBServiceSpecResolver: &db_resolver.ServiceSpecResolver{DB: r.DB, ServiceSpec: serviceSpec}}, nil
+	} else {
+		return nil, fmt.Errorf("Delete the service mapped to this service-spec first.")
```
Comment (aballman) [inline plugins/codeamp/graphql/servicespec_mutation.go]: log

File: plugins/codeamp/graphql/servicespec_query.go [hunk 9]
```diff
@@ -22,7 +22,7 @@ func (r *ServiceSpecResolverQuery) ServiceSpecs(ctx context.Context) ([]*Service
 	var rows []model.ServiceSpec
 	var results []*ServiceSpecResolver
 
-	r.DB.Order("created_at desc").Find(&rows)
+	r.DB.Where("type != ?", "suggested").Order("created_at desc").Find(&rows)
```
Comment (aballman) [inline plugins/codeamp/graphql/servicespec_query.go]: log if not not-found error

File: plugins/smartprofiles/helpers.go [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+	1. memory_current_cost --> telegraf.autogen.kubernetes_pod_container.memory_usage_bytes
+	2. memory_min_cost     --> telegraf.autogen.prom_kube_pod_container_resource_requests_memory_bytes.gauge
+	3. memory_max_cost     --> telegraf.autogen.prom_kube_pod_container_resource_limits_memory_bytes.gauge
+	4. cpu_current_cost    --> telegraf.autogen.kubernetes_pod_container.cpu_usage_nanocores
+	5. cpu_min_cost        --> telegraf.autogen.prom_kube_pod_container_resource_requests_cpu_cores.gauge
+	6. cpu_max_cost        --> telegraf.autogen.prom_kube_pod_container_resource_limits_cpu_cores.gauge
+*/
+
+// CPUUnitCost is the rate per hour to run a cpu on a r4.4xlarge node
+const CPUUnitCost = 0.07
```
Comment (aballman) [inline plugins/smartprofiles/helpers.go]: These should be configurable without a code change.

File: plugins/codeamp/db/service.go (hunk 1 above)
Comment (drshrey) [reply plugins/codeamp/db/service.go]: To return the original service spec associated with the service, not autogenerated ones

File: plugins/codeamp/db/service.go (hunk 2 above)
Comment (drshrey) [reply plugins/codeamp/db/service.go]: 👍

File: plugins/smartprofiles/helpers.go (hunk 10 above)
Comment (drshrey) [reply plugins/smartprofiles/helpers.go]: Agreed, will define as global vars and move at top of smartprofiles.go. If input config doesn't have the two, then I'll return an error to mark them as required

Comment (drshrey) [reply plugins/smartprofiles/helpers.go]: We don't use this anymore actually, so I'm just going to remove it

File: plugins/smartprofiles/interfaces.go [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+		return
+	}
+
+	memRecommendation := GetResourceRecommendation(memoryCost)
+	cpuRecommendation := GetResourceRecommendation(cpuCost)
+
+	memDifference := GetResourceDiff(*memoryCost, *memRecommendation)
+	cpuDifference := GetResourceDiff(*cpuCost, *cpuRecommendation)
+
+	fmt.Println(memRecommendation, cpuRecommendation)
```
Comment (amoskyler) [inline plugins/smartprofiles/interfaces.go]: should use logger

File: plugins/smartprofiles/interfaces.go [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+	return nil
+}
+
+
+// Computes Service details for a given request
+func (ic *SmartProfilesClient) GetService(id string, name string, namespace string, timeRange string, svcChan chan *Service) {
+	fmt.Println(fmt.Sprintf("[...] appending %s - %s", name, namespace))
+	memoryCost, err := GetServiceMemoryCost(ic, name, namespace, timeRange)
+	if err != nil {
+		fmt.Println(err.Error())
```
Comment (amoskyler) [inline plugins/smartprofiles/interfaces.go]: logger

File: plugins/smartprofiles/helpers.go [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+
+	minCostFloat, err := ComputeMeanSampledQuery(ic, fmt.Sprintf("mean(gauge) from prom_kube_pod_container_resource_requests_cpu_cores where container='%s' and namespace='%s'", serviceName, namespace), 14)
+	if err != nil {
+		spew.Dump(err)
+		return nil, err
+	}	
+
+	maxCostFloat, err := ComputeMeanSampledQuery(ic, fmt.Sprintf("mean(gauge) from prom_kube_pod_container_resource_limits_cpu_cores where container='%s' and namespace='%s'", serviceName, namespace), 14)
+	if err != nil {
+		spew.Dump(err)
```
Comment (amoskyler) [inline plugins/smartprofiles/helpers.go]: rm spew

File: plugins/smartprofiles/helpers.go [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+	// cpu_min_cost        --> telegraf.autogen.prom_kube_pod_container_resource_requests_cpu_cores.gauge
+	// cpu_max_cost        --> telegraf.autogen.prom_kube_pod_container_resource_limits_cpu_cores.gauge
+
+	// min > current by atleast 20%
+	overProvisioned := false
+
+	// get current cost
+	currentCostFloat, err := ComputeMeanSampledQuery(ic, fmt.Sprintf("mean(cpu_usage_nanocores)/1000 from kubernetes_pod_container where container_name='%s' and namespace = '%s'", serviceName, namespace), 14)
+	if err != nil {
+		spew.Dump(err)
```
Comment (amoskyler) [inline plugins/smartprofiles/helpers.go]: rm all spew

File: plugins/smartprofiles/helpers.go [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+func GetServiceMemoryCost(ic SmartProfilesClienter, serviceName string, namespace string, timeRange string) (*Resource, error) {
+	// memory_current_cost --> telegraf.autogen.kubernetes_pod_container.memory_usage_bytes
+	// memory_min_cost     --> telegraf.autogen.prom_kube_pod_container_resource_requests_memory_bytes.gauge
+	// memory_max_cost     --> telegraf.autogen.prom_kube_pod_container_resource_limits_memory_bytes.gauge
+
+	// min > current by atleast 20%
+	overProvisioned := false
+
+	// get current cost
+	currentCostFloat, err := ComputeMeanSampledQuery(ic, fmt.Sprintf("mean(memory_usage_bytes)/1000000 from kubernetes_pod_container where container_name='%s' and namespace='%s'", serviceName, namespace), 14)
```
Comment (amoskyler) [inline plugins/smartprofiles/helpers.go]: You should assign `1000000` to a constant with a legible name and replace in all of the places you use the raw int.

Comment (amoskyler) [reply plugins/smartprofiles/helpers.go]: Something like `MEGABYTE` would be reasonable.

File: plugins/smartprofiles/interfaces_test.go [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+func (ic *MockSmartProfilesClient) InitInfluxClient(influxHost string, influxDBName string) (error) {
+	ic.InfluxClient = MockInfluxClient{}
+	ic.InfluxDBName = "telegraf"
+
+	return nil
+}
+
+// GetService
+func (ic *MockSmartProfilesClient) GetService(id string, name string, namespace string, timeRange string, svcChan chan *smartprofiles.Service) {
+	spew.Dump("MockInfluxClient GetService")
```
Comment (amoskyler) [inline plugins/smartprofiles/interfaces_test.go]: spewwwwww

File: plugins/smartprofiles/smartprofiles.go [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+	"github.com/codeamp/transistor"
+	// "github.com/davecgh/go-spew/spew"
+
+	log "github.com/codeamp/logger"
+)
+
+//SmartProfiles is a local struct for smartprofiles plugin
+type SmartProfiles struct {
+	events chan transistor.Event
+	SmartProfilesClienter
```
Comment (amoskyler) [inline plugins/smartprofiles/smartprofiles.go]: what is a Clienter?

File: plugins/smartprofiles/helpers.go [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+	// limit recommendation - 20% above p90
+	p90Float, _ := strconv.ParseFloat(cost.P90, 64)
+	r.Limit = fmt.Sprintf("%f", float64(1.2)*p90Float)
+	r.OverProvisioned = cost.OverProvisioned
+
+	return &r
+}
+
+/*
+ComputeMeanSampledQuery takes in a query and how many days to sample and returns
```
Comment (amoskyler) [inline plugins/smartprofiles/helpers.go]: This function doesn't really sample. It simply calculates the averages of each day one at a time. This should optimally accept a sample rate (in terms of % of time sampled) and return based on that.

**Pull request closed without merging**